Promote files atomically

snowleopard opened the pull request:
This PR makes another step towards more reliable file promotion: it changes Dune to promote files as an atomic operation, to avoid retriggering Dune multiple times while observing partial results as the promoted file is being copied and processed by the artifact substitution logic. 

Perhaps, we should just make `Artifact_substitution.copy_file` atomic instead of adding the second function?

jeremiedimino commented:
> Perhaps, we should just make Artifact_substitution.copy_file atomic instead of adding the second function?

That might not work so well for `dune install`, since we don't have a temp dir at hand that we know will be on the right partition.

jeremiedimino commented on src/fiber/fiber.mli:
```diff
@@ -387,3 +387,20 @@ type fill = Fill : 'a Ivar.t * 'a -> fill
     fill. *)
 val run : 'a t -> iter:(unit -> fill) -> 'a
+
+(** [Temp.Monad] instantiated to the Fiber monad. *)
```
So far we've kept the `Fiber` module completely free of anything IO related. I'm a little sad to depart from this. Could we put this in a separate module/library?

snowleopard replied:
Yes, sure. Does `fiber_util` sound good?

jeremiedimino replied:
Sounds good to me

snowleopard replied:
Done.

snowleopard commented:
@jeremiedimino Good point. OK, then it makes sense to keep both functions.

jeremiedimino reviewed (changes requested):
That seems like an improvement, however it is not going to work when the build directory is on a different partition. Dune allows that, so it might happen. And it seems reasonable to do, for instance one might want to put the build directory on a faster but more volatile medium.

What about using one of the pattern we ignore, such as the temporary editor files? For instance, if we want to create `bin/main.exe`, we use `bin/.#main.exe.temp` as temporary file. Such a pattern would work with `dune install` as well, so we could get rid of the non-atomic version.

snowleopard commented:
> That seems like an improvement, however it is not going to work when the build directory is on a different partition. Dune allows that, so it might happen. And it seems reasonable to do, for instance one might want to put the build directory on a faster but more volatile medium.
> 
> What about using one of the pattern we ignore in the file watcher, such as the temporary editor files? For instance, if we want to create `bin/main.exe`, we use `bin/.#main.exe.temp` as temporary file. Such a pattern would work with `dune install` as well, so we could get rid of the non-atomic version.

Right, I didn't realise the build directory can be on a different partition. Your suggestion sounds good, I'll do that!

snowleopard commented:
I switched to using `.#` temporary files as discussed but for some reason this seem to trigger rebuilds? I though we fixed this recently... Will investigate tomorrow.

snowleopard commented:
Adding `^\.#.*` to the list of excluded patterns fixes the retriggering in the test. This is weird: why do we watch temporary files at the root? I think it's a bug so I'm going to fix this in a separate PR.

snowleopard commented:
@jeremiedimino I addressed your comments, so this PR should be ready to merge.

jeremiedimino reviewed (approved):
Looks good!
